### PR TITLE
Cleanup of Kotlin entity codegen with Schema type hierachy.

### PR DESCRIPTION
**Pull request by piotrswigon**
(Builds up on #5728 which adjusts the goldens to allow this change)

Unlocks  `List<&Thing>` in Kotlin codegen.

Instead of adding another layer of indirection to code that has been been already far too twisted and with too many layers, I instead cleaned it up but introducing a recursive type hierarchy for schema fields.

Fixes b/159856888

### CODE REVIEW

File: src/tools/kotlin-entity-generator.ts [hunk 1]
```diff
@@ -323,8 +282,8 @@ var ${escaped}: ${type}
         }`;
   }
 
-  private mapFields(mapper: (arg: KotlinSchemaField) => string): string[] {
-    return this.entityDescriptor.fields.map(mapper);
+  private mapFields(mapper: (arg: ExtendedSchemaField) => string): string[] {
```
Comment (alxrsngrtn) [inline src/tools/kotlin-entity-generator.ts:285]: You can use the definition from `hot.ts`: `mapper: Mapper<ExtendedSchemaField, string>` if you'd like.

File: src/tools/kotlin-entity-generator.ts [hunk 2]
```diff
@@ -323,8 +282,8 @@ var ${escaped}: ${type}
...
+    return this.fields.map(mapper);
```
Comment (alxrsngrtn) [inline src/tools/kotlin-entity-generator.ts:286]: Now that this function is so simple (no property lookup), is it really needed?

File: src/tools/kotlin-schema-field.ts [hunk 3]
```diff
@@ -0,0 +1,268 @@
...
+ * @license
+ * Copyright (c) 2020 Google Inc. All rights reserved.
+ * This code may only be used under the BSD style license found at
+ * http://polymer.github.io/LICENSE.txt
+ * Code distributed by Google as part of this project is also
+ * subject to an additional IP rights grant found at
+ * http://polymer.github.io/PATENTS.txt
+ */
+
+import {SchemaNode} from './schema2graph.js';
```
Comment (alxrsngrtn) [inline src/tools/kotlin-schema-field.ts:11]: Breaking this out into another file will likely require modification of blaze rules in G3 (depending on if they're using a glob pattern).

File: src/tools/kotlin-schema-field.ts [hunk 4]
```diff
@@ -0,0 +1,268 @@
...
+ * SchemaFieldType
+ *  ├─TypeContainer
+ *  │  ├─CollectionType
+ *  │  ├─ListType
+ *  │  └─ReferenceType
+ *  ├─SchemaType
+ *  │  ├─ReferencedSchema
+ *  │  └─InlineSchema
+ *  └─PrimitiveType
+ */
```
Comment (alxrsngrtn) [inline src/tools/kotlin-schema-field.ts:54]: This is a helpful comment, very nice.

File: src/tools/kotlin-schema-field.ts [hunk 5]
```diff
@@ -0,0 +1,268 @@
...
+    return `emptySet<${this.innerKotlinType}>()`;
+  }
+
+  get decodeFn() {
+    return this.innerType.decodeFn;
+  }
+}
+
+/** List type, e.g. List<Text>, List<inline Object {...}> */
+class ListType extends TypeContainer {
```
Comment (alxrsngrtn) [inline src/tools/kotlin-schema-field.ts:133]: Does list have a `decodeFn`? If not, why not?

File: src/tools/kotlin-schema-field.ts [hunk 6]
```diff
@@ -0,0 +1,268 @@
...
+  constructor(innerType: SchemaFieldType) {
+    super(innerType);
+  }
+
+  getKotlinType() {
+    return `List<${this.innerKotlinType}>`;
+  }
+
+  get defaultVal() {
+    return `listOf<${this.innerKotlinType}>()`;
```
Comment (alxrsngrtn) [inline src/tools/kotlin-schema-field.ts:144]: `emptyList()` -- unless, do we need to specify the type information?

File: src/tools/tests/kotlin-schema-field-test.ts [hunk 7]
```diff
@@ -0,0 +1,68 @@
...
+ */
+import {assert} from '../../platform/chai-web.js';
+import {Manifest} from '../../runtime/manifest.js';
+import {SchemaGraph} from '../schema2graph.js';
+import {generateFields} from '../kotlin-schema-field.js';
+
+describe('Kotlin Schema Fields', () => {
+  it('generates Kotlin Types', async () => {
+
+    const expectations: [string, string][] = [
```
Comment (alxrsngrtn) [inline src/tools/tests/kotlin-schema-field-test.ts:18]: Awesome! I'm glad that there are tests for the new file.

File: src/tools/kotlin-schema-field.ts (hunk 5 above)
Comment (piotrswigon) [reply src/tools/kotlin-schema-field.ts:133]: decodeFns turn out to only be used in WASM, and WASM doesn't have support for any of these more complex types.

File: src/tools/kotlin-schema-field.ts (hunk 6 above)
Comment (piotrswigon) [reply src/tools/kotlin-schema-field.ts:144]: I doesn't seem like we have to given the type explicitly, updating.

File: src/tools/tests/kotlin-schema-field-test.ts (hunk 7 above)
Comment (piotrswigon) [reply src/tools/tests/kotlin-schema-field-test.ts:18]: They are limited, but still better than what came before.

File: src/tools/kotlin-schema-field.ts (hunk 3 above)
Comment (piotrswigon) [reply src/tools/kotlin-schema-field.ts:11]: I've set it up a number of weeks ago with globs() so you can add and remove files without needing to adjust the BUILD file. Yay!

File: src/tools/kotlin-entity-generator.ts (hunk 2 above)
Comment (piotrswigon) [reply src/tools/kotlin-entity-generator.ts:286]: Indeed! Thanks! s/mapField/fields.map/ woohoo!

**Comment by piotrswigon** [comment]
Yup, agreed @cromwellian. This should be moved to visitor pattern the next time we want to use this hierarchy for anything. Currently it is only used for Kotlin type generation, but it is clear that we use schema field types in a lot of places and we should lift this hierachy here and add a visitor pattern... It would also clean up the C++ code FWIW.

But well... one step at a time.

**Pull request merged**